src/mds/Server.cc

```diff
@@ -6407,6 +6407,12 @@ void Server::handle_client_setxattr(MDRequestRef& mdr)
   size_t len = req->get_data().length();
   size_t inc = len + name.length();
 
+  if (inc > mds->mdsmap->get_max_xattr_size()) {
+    dout(10) << "New xattr pair too big: " << inc << dendl;
+    respond_to_request(mdr, -CEPHFS_ENOSPC);
+    return;
+  }
+
```

Comment:
I think you can just discard the above change.

And move the `size_t cur_xattrs_size = 0;` outside of and before the `if(pxattrs) {}` scope and then 
simplify it and move  https://github.com/ceph/ceph/pull/46357/commits/3effc1758deeeaee038b48939e4dfa7f6f2c3d6d#diff-277dc6e796ccecb6aa14c9357f7a86898d6ddcf4113c110a4e00e0a10f6fefa7R6428-R6433 outside and just after `if(pxattrs) {}` ?

Reply:
Maybe it's because it's Friday (or simply because my brain melted...), but I can't understand your comment.  This check is to prevent exceeding the max size when setting the **first** xattr in an inode.  Can you please clarify?

Reply:
Before you set the first xattr, doesn't the `pxattrs` equal to `nulltr` ? Then the `cur_xattrs_size` will be 0.

Reply:
Sure, it will be NULL, but that's not the point.  Even if the `if (pxattrs) { ... }` block isn't executed (which happens when you're setting the first  xattr), you'll still set the new xattr **without** checking it's size.  And that's what this extra commit in this PR does.

Just mount a filesystem and run:
```
dd if=/dev/zero bs=1 count=65536 2> /dev/null | attr -s my-attr-name file
```
(You'll probably need to make sure you don't have `Xx` caps, otherwise you may hit the other size checks in this PR.)

Reply:
I mean:

```
diff --git a/src/mds/Server.cc b/src/mds/Server.cc
index e8313712cdc..e86df249269 100644
--- a/src/mds/Server.cc
+++ b/src/mds/Server.cc
@@ -6409,22 +6409,22 @@ void Server::handle_client_setxattr(MDRequestRef& mdr)
 
   auto handler = Server::get_xattr_or_default_handler(name);
   const auto& pxattrs = cur->get_projected_xattrs();
+  size_t cur_xattrs_size = 0;
   if (pxattrs) {
     // check xattrs kv pairs size
-    size_t cur_xattrs_size = 0;
     for (const auto& p : *pxattrs) {
       if ((flags & CEPH_XATTR_REPLACE) && name.compare(p.first) == 0) {
        continue;
       }
       cur_xattrs_size += p.first.length() + p.second.length();
     }
+  }
 
-    if (((cur_xattrs_size + inc) > g_conf()->mds_max_xattr_pairs_size)) {
-      dout(10) << "xattr kv pairs size too big. cur_xattrs_size "
-       << cur_xattrs_size << ", inc " << inc << dendl;
-      respond_to_request(mdr, -CEPHFS_ENOSPC);
-      return;
-    }
+  if (((cur_xattrs_size + inc) > g_conf()->mds_max_xattr_pairs_size)) {
+    dout(10) << "xattr kv pairs size too big. cur_xattrs_size "
+      << cur_xattrs_size << ", inc " << inc << dendl;
+    respond_to_request(mdr, -CEPHFS_ENOSPC);
+    return;
   }
```
Won't this work ?

Reply:
Ah, got it!  OK, that makes sense indeed.  I'll update that commit to simplify the code as you suggest.  Thanks @lxbsz